tools/autograd/templates/variable_factories.h

```diff
@@ -20,6 +21,134 @@ using at::DimnameList;
 
 namespace torch {
 
+namespace detail {
+  enum class ListInitTensorType { Scalar, InitList };
+
+  // We use `ListInitTensor` to support converting an arbitrarily nested braced-init-list
+  // (e.g. {{1, 2}, {3, 4}}) into the equivalent Tensor, taking advantage of the fact that
+  // the constructor will automatically be called recursively until it reaches all innermost
+  // scalar values.
+  //
+  // At any time, a `ListInitTensor` object represents either of the following:
+  // 1. A scalar with value `scalar()` and type `scalar_type()`.
+  // 2. A Tensor represented in `std::initializer_list<ListInitTensor>` form, with value
+  //    `init_list()`, Tensor scalar type `scalar_type()`, and Tensor sizes `sizes()`.
+  struct ListInitTensor {
```

Comment:
I don't get why this is in variable_factories -- this seems to apply perfectly fine to `at::`.